arch/RISCV/RISCVDisassembler.c

```diff
@@ -0,0 +1,246 @@
+#include "../../MCRegisterInfo.h"
+#include "../../MCDisassembler.h"
+#include "../../MathExtras.h"
+
+#define GET_SUBTARGETINFO_ENUM
+#include "RISCVGenSubtargetInfo.inc"
+
+static uint64_t getFeatureBits(int mode)
+{
+	// support everything
+	return (uint64_t) - 1;
+}
+
+// Forward declare these because the autogenerated code will reference them.
+// Definitions are further down.
+static DecodeStatus DecodeGPRRegisterClass(MCInst * Inst, uint64_t RegNo,
+					   uint64_t Address,
+					   const void *Decoder);
+static DecodeStatus DecodeFPR32RegisterClass(MCInst * Inst, uint64_t RegNo,
+					     uint64_t Address,
+					     const void *Decoder);
+static DecodeStatus DecodeFPR64RegisterClass(MCInst * Inst, uint64_t RegNo,
+					     uint64_t Address,
+					     const void *Decoder);
+static DecodeStatus decodeUImmOperand(MCInst * Inst, uint64_t Imm,
+				      int64_t Address, const void *Decoder,
+				      unsigned N);
+static DecodeStatus decodeSImmOperand(MCInst * Inst, uint64_t Imm,
+				      int64_t Address, const void *Decoder,
+				      unsigned N);
+static DecodeStatus decodeSImmOperandAndLsl1(MCInst * Inst, uint64_t Imm,
+					     int64_t Address,
+					     const void *Decoder, unsigned N);
+
+#include "RISCVGenDisassemblerTables.inc"
+
+#define GET_REGINFO_ENUM
+#include "RISCVGenRegisterInfo.inc"
+
+#define GET_REGINFO_MC_DESC
+#include "RISCVGenRegisterInfo.inc"
+
+#define GET_INSTRINFO_ENUM
+#include "RISCVGenInstrInfo.inc"
+
+void RISCV_init(MCRegisterInfo * MRI)
+{
+	/*InitMCRegisterInfo(RISCVRegDesc, 97, RA, PC,
+	   RISCVMCRegisterClasses, 3,
+	   RISCVRegUnitRoots,
+	   64,
+	   RISCVRegDiffLists,
+	   RISCVLaneMaskLists,
+	   RISCVRegStrings,
+	   RISCVRegClassStrings,
+	   RISCVSubRegIdxLists,
+	   2,
+	   RISCVSubRegIdxRanges,
+	   RISCVRegEncodingTable);
+	 */
+
+	MCRegisterInfo_InitMCRegisterInfo(MRI, RISCVRegDesc, 97,
+					  0, 0,
+					  RISCVMCRegisterClasses, 3,
+					  0, 0,
+					  RISCVRegDiffLists,
+					  0, RISCVSubRegIdxLists, 2, 0);
+}
+
+static DecodeStatus RISCVDisassembler_getInstruction(int mode, MCInst * MI,
+				 const uint8_t * code, size_t code_len,
+				 uint64_t * Size, uint64_t Address,
+				 MCRegisterInfo * MRI)
+{
+	/*MCInst &MI, uint64_t &Size,
+	   ArrayRef<uint8_t> Bytes,
+	   uint64_t Address,
+	   raw_ostream &OS,
+	   raw_ostream &CS) const { */
+
+	// TODO: although assuming 4-byte instructions is sufficient for RV32 and
+	// RV64, this will need modification when supporting the compressed
+	// instruction set extension (RVC) which uses 16-bit instructions. Other
+	// instruction set extensions have the option of defining instructions up to
+	// 176 bits wide.
+	*Size = 4;
+	if (code_len < 4) {
+		*Size = 0;
+		return MCDisassembler_Fail;
+	}
+
+	if (MI->flat_insn->detail) {
+		memset(MI->flat_insn->detail, 0, sizeof(cs_detail));
+	}
+	// Get the four bytes of the instruction.
+	//Encoded as little endian 32 bits.
+	uint32_t Inst = (code[0] << 0) |
```

Comment:
please put all the variable declarations at top of code block, not at random places, so older compiler like MSVC2010 still works.

currently we have this issue in many places in this PR.

Reply:
I found two files had this similar issues and all fixed now.